drivers/display/display_stm32_ltdc.c

```diff
@@ -171,58 +175,151 @@ static int stm32_ltdc_write(const struct device *dev, const uint16_t x,
-
 		for (row = 0; row < desc->height; row++) {
 			(void) memcpy(dst, src, desc->width * data->current_pixel_size);
-			sys_cache_data_flush_range(dst, desc->width * data->current_pixel_size);
 			dst += (config->width * data->current_pixel_size);
 			src += (desc->pitch * data->current_pixel_size);
 		}
+		goto swap;
+	}
 
+	if (data->current_orientation != DISPLAY_ORIENTATION_ROTATED_90) {
+		LOG_ERR("Only 90° rotation supported");
+		return -ENOTSUP;
 	}
 
+	/* Perform rotation operations for 90°. Note that the physical display is rotated 90°,
+	 * meaning we are to perform a frame buffer rotation of 270° to make the display look
+	 * right.
+	 */
+
+	dst += (y * data->current_pixel_size);
+	dst += ((config->height - x - desc->width) * config->width * data->current_pixel_size);
+	/* _x, _y = coordinates in src */
+	for (int _x = desc->width - 1; _x > 0; _x--) {
+		/* y_dst = destination for the desc->height amount of vertical
+		 *         pixels in src's _x column to be copied to row y in dst
+		 */
+		uint8_t *y_dst = &dst[(desc->width - _x - 1) * config->width *
+				      data->current_pixel_size];
+		const uint8_t *y_src = &src[_x*data->current_pixel_size];
+
+		for (int _y = 0; _y < desc->height; _y++) {
+			/* For performance reasons, don't call memcpy on such small values.
+			 * E.g. STM32H743 w/ 16-bit SDRAM bus & 1024x600 & 90° rotation:
+			 *   - memcpy: 11fps
+			 *   - this: 15fps
+			 */
+			switch (data->current_pixel_size) {
+			case 1:
+				y_dst[0] = y_src[0];
+				break;
+			case 2:
+				y_dst[0] = y_src[0];
+				y_dst[1] = y_src[1];
+				break;
+			case 4:
+				y_dst[0] = y_src[0];
+				y_dst[1] = y_src[1];
+				y_dst[2] = y_src[2];
+				y_dst[3] = y_src[3];
+				break;
+			}
+			y_dst += data->current_pixel_size;
+			y_src += desc->width * data->current_pixel_size;
+		}
+	}
+swap:
+
+	/* More to come? */
+	if (desc->frame_incomplete) {
+		return 0;
+	}
+
+	sys_cache_data_flush_range(pend_buf, data->frame_buffer_len);
+
 	if (data->front_buf == pend_buf) {
 		return 0;
 	}
 
-	k_sem_reset(&data->sem);
+	/* If device is suspended, don't swap the buffers as the ISR is not running */
+	enum pm_device_state pm_state = PM_DEVICE_STATE_SUSPENDED;
 
-	data->pend_buf = pend_buf;
+	if (pm_device_state_get(dev, &pm_state) == 0 && pm_state == PM_DEVICE_STATE_ACTIVE) {
+		k_sem_reset(&data->sem);
+		data->pend_buf = pend_buf;
+		k_sem_take(&data->sem, K_FOREVER);
+	}
+
+	if (pend_buf == NULL || old == NULL) {
+		return 0;
+	}
 
-	k_sem_take(&data->sem, K_FOREVER);
+	/* Safe to copy the new frame to the old one */
+
+#if DT_INST_NODE_HAS_PROP(0, ext_sdram) && CONFIG_STM32_LTDC_INTERMEDIATE_BUF_SIZE > 0
+	/* Copy into an intermediate buffer to speed up copying,
+	 * as linear access is faster in SDRAM memories, as compared to byte-wise copying
+	 *
+	 * E.g. STM32H743 w/ 16-bit SDRAM bus & 1024x600 & 90° rotation:
+	 *   - memcpy: 7fps
+	 *   - this: 11fps
+	 */
+	static uint8_t tmp[CONFIG_STM32_LTDC_INTERMEDIATE_BUF_SIZE];
```

Comment:
not sure if there can be multiple LTDC instances but this static buffer would be an issue as this makes stm32_ltdc_write non-reentrant